File: docs/aimdk/src/py_examples/py_examples/play_linkcraft.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from aimdk_msgs.srv import GetRobotResources, ExecuteActionResource


class RobotResourceClient(Node):
    def __init__(self):
        super().__init__('robot_resource_client')

        self.get_resource_client = self.create_client(
            GetRobotResources,
            '/aimdk_5Fmsgs/srv/GetRobotResources'
        )
        self.play_resource_client = self.create_client(
            ExecuteActionResource,
            '/aimdk_5Fmsgs/srv/ExecuteActionResource'
        )

        self.resources = []

    def get_robot_resources(self):
        while not self.get_resource_client.wait_for_service(timeout_sec=1.0):
            if not rclpy.ok():
                self.get_logger().error("Interrupted while waiting for the service. Exiting.")
                return False
            self.get_logger().info("Service not available, waiting again...")

        request = GetRobotResources.Request()
        future = self.get_resource_client.call_async(request)
        rclpy.spin_until_future_complete(self, future)

        if future.result() is None:
            self.get_logger().error("Failed to call GetRobotResources")
            return False

        response = future.result()
        self.resources = list(response.robot_resources)
        self.get_logger().info(f"Found {len(self.resources)} resources")
        return True

    def print_resource_menu(self):
        if not self.resources:
            print("(no resources)")
            return

        print("\n=== Available Robot Resources ===")
        for idx, resource in enumerate(self.resources):
            ver = resource.current_version
            print(f"[{idx}] {ver.name or '<no name>'}")
            print(f"     key:     {resource.resource_key}")
            print(f"     version: {ver.version}")
        print("=================================\n")

    def pick_resource(self):
        if not self.resources:
            return None

        while True:
            try:
                raw = input(
                    f"Select resource index [0-{len(self.resources) - 1}] (q to quit): ").strip()
            except EOFError:
                return None
            if raw.lower() in ("q", "quit", "exit"):
                return None
            if not raw.isdigit():
                print("Please enter a number.")
                continue
            idx = int(raw)
            if 0 <= idx < len(self.resources):
                return self.resources[idx]
            print(f"Index out of range (0-{len(self.resources) - 1}).")

    def play_resource(self, resource):
        while not self.play_resource_client.wait_for_service(timeout_sec=1.0):
            if not rclpy.ok():
                self.get_logger().error("Interrupted while waiting for service.")
                return
            self.get_logger().info("Service not available, waiting...")

        request = ExecuteActionResource.Request()
        request.resource_key = resource.resource_key
        request.resource_version = resource.current_version.version

        if "onnx" in request.resource_key:
            request.meta = '{"resource_type": "BODY_MONTION"}'
        else:
            request.meta = '{"resource_type": "ARM_MONTION"}'

        self.get_logger().info("Sending request:")
        self.get_logger().info(f"  resource_key: {request.resource_key}")
        self.get_logger().info(
            f"  resource_version: {request.resource_version}")
        self.get_logger().info(f"  meta: {request.meta}")

        future = self.play_resource_client.call_async(request)
        rclpy.spin_until_future_complete(self, future)

        if future.result() is not None:
            self.handle_response(future.result())
        else:
            self.get_logger().error("Failed to call ExecuteActionResource")

    def handle_response(self, response):
        self.get_logger().info("Service response received:")
        self.get_logger().info("Response Header:")
        self.get_logger().info(
            f"  stamp: {response.header.header.stamp.sec}.{response.header.header.stamp.nanosec:09}")
        self.get_logger().info(f"  code: {response.header.header.code}")
        self.get_logger().info("Status:")
        self.get_logger().info(f"  value: {response.header.status.value}")
        self.get_logger().info(f"  message: {response.header.message}")


def main(args=None):
    rclpy.init(args=args)
    client_node = None
    try:
        client_node = RobotResourceClient()
        if not client_node.get_robot_resources():
            return
        client_node.print_resource_menu()
        chosen = client_node.pick_resource()
        if chosen is None:
            client_node.get_logger().info("No resource selected, exiting.")
            return
        client_node.play_resource(chosen)
    except KeyboardInterrupt:
        pass
    finally:
        if client_node:
            client_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: docs/aimdk/src/examples/src/interaction/play_linkcraft.cpp
//
// Created by agiuser on 2026/3/3.
//

#include <aimdk_msgs/srv/execute_action_resource.hpp>
#include <aimdk_msgs/srv/get_robot_resources.hpp>
#include <iostream>
#include <rclcpp/rclcpp.hpp>
#include <string>
#include <vector>

using namespace std::chrono_literals;

class RobotResourceClient : public rclcpp::Node {
public:
  RobotResourceClient() : Node("robot_resource_client") {
    get_resource_client_ =
        this->create_client<aimdk_msgs::srv::GetRobotResources>(
            "/aimdk_5Fmsgs/srv/GetRobotResources");
    play_resource_client_ =
        this->create_client<aimdk_msgs::srv::ExecuteActionResource>(
            "/aimdk_5Fmsgs/srv/ExecuteActionResource");
  }

  bool get_robot_resources() {
    while (!get_resource_client_->wait_for_service(1s)) {
      if (!rclcpp::ok()) {
        return false;
      }
      RCLCPP_INFO(this->get_logger(),
                  "Service not available, waiting again...");
    }

    if (!rclcpp::ok()) {
      return false;
    }

    auto request =
        std::make_shared<aimdk_msgs::srv::GetRobotResources::Request>();

    auto future_result = get_resource_client_->async_send_request(request);

    if (rclcpp::spin_until_future_complete(this->get_node_base_interface(),
                                           future_result) !=
        rclcpp::FutureReturnCode::SUCCESS) {
      if (rclcpp::ok()) {
        RCLCPP_ERROR(this->get_logger(), "Failed to call GetRobotResources");
      }
      return false;
    }

    auto response = future_result.get();
    resources_ = response->robot_resources;
    RCLCPP_INFO(this->get_logger(), "Found %zu resources", resources_.size());
    return true;
  }

  void print_resource_menu() {
    if (resources_.empty()) {
      std::cout << "(no resources)\n";
      return;
    }

    std::cout << "\n=== Available Robot Resources ===\n";
    for (size_t i = 0; i < resources_.size(); ++i) {
      const auto &ver = resources_[i].current_version;
      std::cout << "[" << i << "] "
                << (ver.name.empty() ? "<no name>" : ver.name) << "\n";
      std::cout << "     key:     " << resources_[i].resource_key << "\n";
      std::cout << "     version: " << ver.version << "\n";
    }
    std::cout << "=================================\n\n";
  }

  int pick_resource() {
    if (resources_.empty()) {
      return -1;
    }

    std::string line;
    while (true) {
      std::cout << "Select resource index [0-" << resources_.size() - 1
                << "] (q to quit): ";
      if (!std::getline(std::cin, line)) {
        return -1;
      }
      if (line == "q" || line == "quit" || line == "exit") {
        return -1;
      }
      try {
        int idx = std::stoi(line);
        if (idx >= 0 && idx < static_cast<int>(resources_.size())) {
          return idx;
        }
        std::cout << "Index out of range (0-" << resources_.size() - 1
                  << ").\n";
      } catch (...) {
        std::cout << "Please enter a number.\n";
      }
    }
  }

  void play_resource(int idx) {
    while (!play_resource_client_->wait_for_service(1s)) {
      if (!rclcpp::ok()) {
        RCLCPP_ERROR(this->get_logger(),
                     "Interrupted while waiting for service.");
        return;
      }
      RCLCPP_INFO(this->get_logger(), "Service not available, waiting...");
    }

    if (!rclcpp::ok()) {
      return;
    }

    const auto &resource = resources_[idx];
    auto request =
        std::make_shared<aimdk_msgs::srv::ExecuteActionResource::Request>();

    request->resource_key = resource.resource_key;
    request->resource_version = resource.current_version.version;

    if (request->resource_key.find("onnx") != std::string::npos) {
      request->meta = R"({"resource_type": "BODY_MONTION"})";
    } else {
      request->meta = R"({"resource_type": "ARM_MONTION"})";
    }

    RCLCPP_INFO(this->get_logger(), "Sending request:");
    RCLCPP_INFO(this->get_logger(), "  resource_key: %s",
                request->resource_key.c_str());
    RCLCPP_INFO(this->get_logger(), "  resource_version: %s",
                request->resource_version.c_str());
    RCLCPP_INFO(this->get_logger(), "  meta: %s", request->meta.c_str());

    auto future = play_resource_client_->async_send_request(request);

    if (rclcpp::spin_until_future_complete(this->get_node_base_interface(),
                                           future) ==
        rclcpp::FutureReturnCode::SUCCESS) {
      handle_response(future.get());
    } else {
      if (rclcpp::ok()) {
        RCLCPP_ERROR(this->get_logger(),
                     "Failed to call ExecuteActionResource");
      }
    }
  }

private:
  rclcpp::Client<aimdk_msgs::srv::GetRobotResources>::SharedPtr
      get_resource_client_;
  rclcpp::Client<aimdk_msgs::srv::ExecuteActionResource>::SharedPtr
      play_resource_client_;
  std::vector<aimdk_msgs::msg::RobotResource> resources_;

  void handle_response(
      const aimdk_msgs::srv::ExecuteActionResource::Response::SharedPtr
          response) {
    RCLCPP_INFO(this->get_logger(), "Service response received:");
    RCLCPP_INFO(this->get_logger(), "Response Header:");
    RCLCPP_INFO(this->get_logger(), "  stamp: %d.%09u",
                response->header.header.stamp.sec,
                response->header.header.stamp.nanosec);
    RCLCPP_INFO(this->get_logger(), "  code: %ld",
                response->header.header.code);
    RCLCPP_INFO(this->get_logger(), "Status:");
    RCLCPP_INFO(this->get_logger(), "  value: %d",
                response->header.status.value);
    RCLCPP_INFO(this->get_logger(), "  message: %s",
                response->header.message.c_str());
  }
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto client_node = std::make_shared<RobotResourceClient>();

  if (!client_node->get_robot_resources()) {
    rclcpp::shutdown();
    return 1;
  }

  client_node->print_resource_menu();
  int chosen = client_node->pick_resource();
  if (chosen < 0) {
    RCLCPP_INFO(client_node->get_logger(), "No resource selected, exiting.");
    rclcpp::shutdown();
    return 0;
  }

  client_node->play_resource(chosen);
  rclcpp::shutdown();
  return 0;
}
